server/stats.go

```diff
@@ -0,0 +1,228 @@
+)
+
+const statsAutosaveInterval = 10 * time.Minute
+const statsAutosaveMaxDither = 60 // seconds
+
+var initStatsOnce sync.Once
+
+func (p *Plugin) initStats() {
+	conf := p.getConfig()
+	if conf.DisableStats || conf.stats != nil {
+		return
+	}
+
+	data, appErr := p.API.KVGet(statsKeyName())
+	if appErr != nil {
+		return
+	}
+	stats := expvar.NewStats(data)
+
+	p.updateConfig(func(c *config) {
+		initStatsOnce.Do(func() {
+			if !c.DisableStats {
+				c.stats = stats
+			}
+		})
+	})
+
+	initUserCounter(p.currentInstanceStore, p.userStore)
+	initUptime()
+
+	p.startAutosaveStats()
+}
+
+func httpAPIStats(p *Plugin, w http.ResponseWriter, r *http.Request) (int, error) {
+	if r.Method != http.MethodGet {
+		return http.StatusMethodNotAllowed,
+			errors.New("method " + r.Method + " is not allowed, must be GET")
+	}
+
+	conf := p.getConfig()
+
+	//TODO protect from unauthorized access?
+	// status, err := verifyWebhookRequestSecret(conf, r)
+	// if err != nil {
+	// 	return status, err
+	// }
+
+	if conf.stats == nil {
+		return http.StatusNotFound, errors.New("No stats available")
+	}
+
+	out := "{"
+	first := true
+	conf.stats.Do(func(name string, e *expvar.Endpoint) {
+		if first {
+			first = false
+		} else {
+			out += ","
+		}
+		out += `"` + name + `":` + e.String()
+	})
+	out += "}"
+	_, err := w.Write([]byte(out))
+	if err != nil {
+		return http.StatusInternalServerError, errors.WithMessage(err, "failed to write response")
+	}
+	return http.StatusOK, nil
+}
+
+// To save the stats periodically, use `go Autosave(...)``
+func (p *Plugin) startAutosaveStats() {
+	stop := make(chan bool)
+	go func() {
+		r := rand.New(rand.NewSource(time.Now().UnixNano()))
+		dither := time.Duration(r.Intn(statsAutosaveMaxDither)) * time.Second
+		time.Sleep(dither)
+
+		ticker := time.NewTicker(statsAutosaveInterval)
+		for {
+			select {
+			case _ = <-stop:
+				return
+			case _ = <-ticker.C:
+				p.saveStats()
+			}
+		}
+	}()
+
+	p.updateConfig(func(c *config) {
+		c.statsStopAutosave = stop
+	})
+}
+
+// To save the stats periodically, use `go Autosave(...)``
+func (p *Plugin) stopAutosaveStats(conf config) {
+	stop := conf.statsStopAutosave
+	if stop == nil {
+		return
+	}
+	p.updateConfig(func(c *config) {
+		c.statsStopAutosave = nil
+	})
+	stop <- true
+}
+
+func (p *Plugin) saveStats() error {
+	stats := p.getConfig().stats
+	if stats == nil {
+		return nil
+	}
+	data, err := json.Marshal(stats)
+	if err != nil {
+		return err
+	}
+	appErr := p.API.KVSet(statsKeyName(), data)
+	if appErr != nil {
+		return appErr
+	}
+	return nil
+}
+
+// This is only useful in a single-server context, so can not be used in production
+// TODO: Need a way to reset all stats in production?
+func (p *Plugin) debugResetStats() error {
+	conf := p.getConfig()
+	p.stopAutosaveStats(conf)
+	defer p.startAutosaveStats()
+
+	stats := conf.stats
+	if stats == nil {
+		return nil
+	}
+	stats.Reset()
+
+	for i := 0; ; i++ {
+		keys, appErr := p.API.KVList(i, listPerPage)
```

Comment:
Maybe an idea for a helper? `foreach` or `range` over all keys. This solution for paging is awesome, but error prone if we have to write it each time.

Reply:
https://github.com/mattermost/mattermost-server/issues/12936